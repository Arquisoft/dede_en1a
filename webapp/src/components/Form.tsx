import { FormEvent, useContext, useState } from "react";
import toast, { Toaster } from "react-hot-toast";
import { CartContext } from "../context/CartContext";
import postData from "../helpers/postData";
import useForm from "../hooks/useForm";
import {Customer, Order} from "../shared/shareddtypes";
import getAddress from "../helpers/getAddress";

const initialState = {
    name: '',
    lastName: '',
    email: '',
    address: ''
}

const notify = (msj: string) => toast(msj);

const Form = () => {

    const {cartItems, dispatch } = useContext(CartContext);
    const {name, email, lastName, address, handleInputChange, resetValues } = useForm<Customer>(initialState);
    const [showToast, setShowToast ] = useState(false);

    const handleSubmit = async (e: FormEvent) => {
        e.preventDefault();

        setShowToast(true);

        const orderDetails = cartItems.map(({id, image, ...item}) => item);

        if(orderDetails.length > 0){

            const order: Order = {
                customer: {
                    name, email, lastName, address
                },
                order_details: orderDetails
            }

            const fetchApi = await postData(order);

            if(!fetchApi.ok){
                notify('An error has occured... You can try again');
            }else{
                notify('Order placed correctly!');

                resetValues();

                dispatch({
                    payload: [],
                    type: 'CLEAR'
                });

            }
        }else {
            notify('The order is empty');
        }

        setTimeout(() => setShowToast(false),5000);
    }


    return (
        <div className='col-md-7 col-lg-8'>
            <h4 className='mb-3'>Checkout</h4>
            <form autoComplete='off' onSubmit={ handleSubmit }>
                <div className="row g-3">
                    <div className="col-sm-6">
                        <label htmlFor="name" className='form-label'>Name</label>
                        <input type="text" className='form-control' name='name' id='name' placeholder='Name..' value={name} onChange={handleInputChange}/>
                    </div>
                    <div className="col-sm-6">
                        <label htmlFor="lastName" className='form-label'>Surnames</label>
                        <input type="text" className='form-control' name='lastName' id='lastName' placeholder='Surname..' value={lastName} onChange={handleInputChange}/>
                    </div>
                    <div className="col-12">
                        <label htmlFor="email" className='form-label'>Email</label>
                        <input type="email" className='form-control' name='email' id='email' placeholder='Email..' value={email} onChange={handleInputChange}/>
                    </div>
                    // TO DO
                    <div className="col-12">
                        <label htmlFor="address" className='form-label'>Address</label>
                        <input type="text" className='form-control' name='address' id='address' placeholder='Address..' value={getAddress("id") + ""}/>
                    </div>
                </div>
                <br />
                <button className='w-100 btn btn-primary' type='submit'>Place order</button>
                {
                    showToast && <Toaster
                        toastOptions={{
                            position: 'top-right',
                            duration: 2000,
                        }}/>
                }
            </form>
        </div>
    )
}

export default Form;