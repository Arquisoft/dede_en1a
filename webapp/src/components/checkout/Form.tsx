import { FormEvent, useContext, useState, useEffect, ChangeEvent } from "react";
import toast, { Toaster } from "react-hot-toast";
import { CartContext } from "../../context/CartContext";
import postData from "../../helpers/postData";
import useForm from "../../hooks/useForm";
import {ContactData, Customer, OrderAdd} from "../../shared/shareddtypes";
import axios from "axios";
import {useSession} from "@inrupt/solid-ui-react";
import {SolidNameComponent} from "../solid/SolidNameComponent";

const initialState = {
    name: '',
    surname: '',
    address: ''
}

const notify = (msj: string) => toast(msj);

function encrypt(webId: string): string {
    return Buffer.from(webId).toString("base64")
}

type Props = {
    setNewAddress: (address: string) => void
}

const Form = (props: Props) => {

    const {setNewAddress} = props
    const {cartItems, dispatch } = useContext(CartContext);
    const {name, surname, address, resetValues } = useForm<Customer>(initialState);
    const [showToast, setShowToast ] = useState(false);
    const [contactData, setContactData] = useState<ContactData[]>();
    const [, setValidValue] = useState(false);

    const {session} = useSession()

    useEffect(() => {
        if(session.info.webId) {
            axios.get((process.env.RESTAPI_URI || "http://localhost:5000") + "/solid/fetch/" + encrypt(session.info.webId)).then(
                response => {
                    localStorage.setItem("fn", response.data[0].fn)
                    setContactData(response.data)
                }
            )
        }
    }, [address, session.info.webId])

    const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
        let selected = e.target.value
        if(selected.length === 0) {
            setValidValue(false)
            setNewAddress("")
            localStorage.removeItem("address")
        } else {
            setValidValue(true)
            setNewAddress(selected)
            localStorage.setItem("address", selected.split("/").join(""))
        }
    }


    const handleSubmit = async (e: FormEvent) => {
        e.preventDefault();

        setShowToast(true);

        const orderDetails = cartItems.map(({_id, image, ...item}) => item);

        if(orderDetails.length > 0){

            const order: OrderAdd = {
                address: localStorage.getItem("address") + "",
                name: localStorage.getItem("name") + "",
                webId: localStorage.getItem("webID") + "",
                items: orderDetails,
                shipping: Number.parseFloat(localStorage.getItem("shipping") + ""),
                totalPrice: Number.parseFloat(localStorage.getItem("totalPrice") + ""),
            }

            const fetchApi = await postData(order);

            if(!fetchApi.ok){
                notify('An error has occurred, try again');
            }else{
                notify('Order placed correctly!');

                resetValues();

                dispatch({
                    payload: [],
                    type: 'CLEAR'
                });

            }
        }else {
            notify('We cannot place an empty order');
        }

        setTimeout(() => setShowToast(false),5000);
    }


    return (
        <div className='col-md-7 col-lg-8'>
            <h4 className='mb-3'>Checkout</h4>
            <form autoComplete='off' onSubmit={ handleSubmit }>
                <div className="row g-3">
                    <div className="col-sm-6">
                        <label htmlFor="name" className='form-label'>
                            Name: <SolidNameComponent/>
                        </label>
                    </div>
                    <div className="col-sm-6">
                        <label htmlFor="surname" className='form-label'>
                            Surname:
                            <input type="text" placeholder="Write your surname..."/>
                        </label>
                    </div>
                    <div>
                        <select name="addressDropdown" id="addressDropdown" onChange={(e) => handleChange(e)}>
                            <option value="">-- Select an address --</option>
                            {contactData?.map((contact, index) => (
                                <option key={index} value={`${contact.country} ${contact.region} ${contact.locality}`}>
                                    {contact.country} / {contact.region} / {contact.locality} / {contact.street_address} / {contact.postal_code}
                                </option>))}
                        </select>
                    </div>
                </div>
                <br />
                <button className='w-100 btn btn-primary' type='submit' onSubmit={handleSubmit}>Place order</button>
                {
                    showToast && <Toaster
                        toastOptions={{
                            position: 'top-right',
                            duration: 2000,
                        }}/>
                }
            </form>
        </div>
    )
}

export default Form;