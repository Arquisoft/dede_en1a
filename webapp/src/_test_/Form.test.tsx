import React from 'react';
import ReactDOM from 'react-dom';
import Router from "react-router-dom";
import { render, screen, cleanup } from '@testing-library/react';
import Form from '../components/checkout/Form';

const setAddress = jest.fn(() => null)

test("Form matches snapshot", () => {
	//TODO: fix this
//   const fakeForm = render(<Form setNewAddress={setAddress} />);
//   expect(fakeForm).toMatchSnapshot();
})

//Test handle change

//Test handle submit