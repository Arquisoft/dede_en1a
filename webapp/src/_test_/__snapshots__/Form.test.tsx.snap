// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="col-md-7 col-lg-8"
      >
        <h4
          class="mb-3"
        >
          Checkout
        </h4>
        <form
          autocomplete="off"
        >
          <div
            class="row g-3"
          >
            <div
              class="col-sm-6"
            >
              <label
                class="form-label"
                for="name"
              >
                Name: 
              </label>
            </div>
            <div>
              <select
                id="addressDropdown"
                name="addressDropdown"
              >
                <option
                  value=""
                >
                  -- Select an address --
                </option>
              </select>
            </div>
          </div>
          <br />
          <button
            class="w-100 btn btn-primary"
            type="submit"
          >
            Place order
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="col-md-7 col-lg-8"
    >
      <h4
        class="mb-3"
      >
        Checkout
      </h4>
      <form
        autocomplete="off"
      >
        <div
          class="row g-3"
        >
          <div
            class="col-sm-6"
          >
            <label
              class="form-label"
              for="name"
            >
              Name: 
            </label>
          </div>
          <div>
            <select
              id="addressDropdown"
              name="addressDropdown"
            >
              <option
                value=""
              >
                -- Select an address --
              </option>
            </select>
          </div>
        </div>
        <br />
        <button
          class="w-100 btn btn-primary"
          type="submit"
        >
          Place order
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
